### PR TITLE
Build a list of dependent constants to recompute in each instance of a generic.

### PR DESCRIPTION
**Pull request by zygoloid**
For each generic, build a list of instructions describing the computations we need to do when resolving an instance of the generic: this is a list of the instance-specific constants and types that the generic uses. Another way of viewing this list is as a block of Carbon SemIR code that is evaluated in order to form an instance of the generic -- this is referenced in the code as the "eval block" for the generic.

For each instruction in the generic whose type or value is a symbolic constant, replace that type or constant value with a symbolic reference that says "to find the actual type or value, look at index N in the list of values for the generic instance".

For an instruction with a symbolic constant value, we can just add that instruction to our list. For an instruction with a symbolic constant type, however, we may not have a corresponding instruction computing the type within the generic and may need to build a new instruction, but will reuse one where possible. In the case where we build a new instruction, we use the existing substitution code to build the type within the eval block.

For now, this transformation is only done in the declaration region of the generic, not in the definition region. Also, we map back from the symbolic references to the underlying constant value in a few places where we will eventually need to do a lookup into a generic instance, in order to avoid regressing the tests.

### CODE REVIEW

File: toolchain/sem_ir/type.h [hunk 1]
```diff
@@ -70,6 +70,13 @@ class TypeStore : public ValueStore<TypeId> {
     return GetAsInst(type_id).TryAs<InstT>();
   }
 
+  // Returns whether two type IDs represent the same type. This includes the
+  // case where they might be in different generics and thus might have
+  // different ConstantIds, but are still symbolically equal.
+  auto EqualAcrossDeclarations(TypeId a, TypeId b) const -> bool {
```
Comment (jonmeow) [inline toolchain/sem_ir/type.h]: FWIW, I'd lean towards IsEqual... or AreEqual... here, similar to IsComplete (for verb phrasing)

File: toolchain/check/subst.h [hunk 2]
```diff
@@ -5,6 +5,7 @@
 #ifndef CARBON_TOOLCHAIN_CHECK_SUBST_H_
 #define CARBON_TOOLCHAIN_CHECK_SUBST_H_
 
+#include "common/map.h"
```
Comment (jonmeow) [inline toolchain/check/subst.h]: I may be missing something, but why is this added? Maybe leftover while refactoring? Perhaps ironically to me, things like llvm::function_ref maybe _should_ have an include.

```suggestion
#include "llvm/ADT/STLFunctionalExtras.h"
```

File: toolchain/check/subst.cpp [hunk 3]
```diff
@@ -106,8 +110,13 @@ static auto ExpandOperands(Context& context, Worklist& worklist,
 static auto PopOperand(Context& context, Worklist& worklist, SemIR::IdKind kind,
                        int32_t arg) -> int32_t {
   switch (kind) {
-    case SemIR::IdKind::For<SemIR::InstId>:
+    case SemIR::IdKind::For<SemIR::InstId>: {
+      auto inst_id = SemIR::InstId(arg);
+      if (!inst_id.is_valid()) {
```
Comment (jonmeow) [inline toolchain/check/subst.cpp:115]: I'll just note, this looks kind of weird versus line 66 in the delta. I don't know if it's worth switching formats since this is consistent with line 122 below, or you could switch both. Or don't do anything, it's probably not going to look as odd later in code, versus when I'm just seeing the delta.

File: toolchain/check/subst.cpp [hunk 4]
```diff
@@ -156,8 +165,8 @@ static auto PopOperand(Context& context, Worklist& worklist, SemIR::IdKind kind,
 
 // Pops the operands of the specified instruction off the worklist and rebuilds
 // the instruction with the updated operands.
```
Comment (jonmeow) [inline toolchain/check/subst.cpp]: This sounds like it would unconditionally call rebuild_fn: should the documentation be adjusted?

File: toolchain/check/subst.h [hunk 5]
```diff
@@ -20,6 +21,24 @@ struct Substitution {
 
 using Substitutions = llvm::ArrayRef<Substitution>;
 
+// A function that performs any needed substitution into an instruction, as part
+// of substituting into a symbolic instruction sequence. The instruction ID
+// should be updated as necessary to represent the new instruction. If the
+// function returns true, the instruction is treated as fully-substituted; if it
+// returns false, the instruction will be decomposed and substitution will be
+// performed recursively into its operands.
```
Comment (jonmeow) [inline toolchain/check/subst.h]: Most of the comments seem to be describing implementation constraints. From reading the implementation, I think this is different: is this a constraint on the caller, or is it a constraint on the implementation? If it's a constraint on the caller, perhaps it should be on SubstInst instead?

File: toolchain/check/subst.h [hunk 6]
```diff
@@ -20,6 +21,24 @@ struct Substitution {
...
+using SubstInstFn = llvm::function_ref<auto(SemIR::InstId& inst_id)->bool>;
+
+// A function that rebuilds an instruction after substitution. `orig_inst_id` is
+// the instruction prior to substitution, and `new_inst` is the substituted
+// instruction. Returns the new instruction ID to use to refer to `new_inst`.
+using SubstRebuildFn = llvm::function_ref<
+    auto(SemIR::InstId orig_inst_id, SemIR::Inst new_inst)->SemIR::InstId>;
+
+// Performs substitution into `inst_id` and its operands recursively.
```
Comment (jonmeow) [inline toolchain/check/subst.h]: Maybe the flow should have some more comments here? From the comments:

subst_fn: "A function that performs any needed substitution into an instruction"
rebuild_fn: "A function that rebuilds an instruction after substitution."
SubstInst: "Performs substitution into `inst_id` and its operands recursively."

- Is `subst_fn` called on all the operands of `inst_id` unconditionally [unconditional on direct, conditional on indirect], or only recursively [conditional on both direct and indirect]?
- On line 237, "// No need to rebuild this instruction.", why do instructions that return false on subst_fn but have no operands not get rebuilt? (or, why are they allowed to return false when they don't need substitution?)
- I've also commented on `Rebuild`, because the call to `rebuild_fn` looks conditional in a way I don't see explained here. [but note since that's going to be in the cpp file, may still be worth a note here]

Here, maybe something like:

> Performs recursive substitution on instructions, starting with `inst_id`. Each instruction is passed to `subst_fn`:
> - On true, the instruction is a leaf.
> - On false, this recurses into the instruction's operands. After all operands are processed, then when [Rebuild condition], the instruction is passed to `rebuild_fn`.

But it may be worth looking at further than just that.

File: toolchain/check/generic.cpp [hunk 7]
```diff
@@ -23,6 +26,124 @@ auto StartGenericDefinition(Context& context) -> void {
...
+    -> SemIR::InstBlockId {
+  context.inst_block_stack().Push();
+
+  Map<SemIR::InstId, SemIR::InstId> constants_in_generic;
+  // TODO: For the definition region, populate constants from the declaration.
+
+  // TODO: See if we can ensure that the generic region stack is unchanged
+  // throughout this work. If so, we can use a range-based loop here instead.
+  for (size_t i = 0;
+       i != context.generic_region_stack().PeekDependentInsts().size(); ++i) {
```
Comment (jonmeow) [inline toolchain/check/generic.cpp]:
```suggestion
  for (auto i : llvm::seq(context.generic_region_stack().PeekDependentInsts().size())) {
```

**Review by jonmeow — APPROVED** [review]
The implementation looks good, but I spent a lot of time examining the substitute/rebuild flow changes due to the lambdas. The flow seems fine (although may be worth considering a class with virtual functions instead of lambda parameters?), but a pass on comments seems like it could be helpful.

File: toolchain/check/subst.cpp (hunk 4 above)
Comment (zygoloid) [reply toolchain/check/subst.cpp]: Done.

File: toolchain/check/subst.h (hunk 2 above)
Comment (zygoloid) [reply toolchain/check/subst.h]: Yeah, left over from refactoring. Just removing the include since I'm also going to look at switching from lambdas to virtual functions.

File: toolchain/sem_ir/type.h (hunk 1 above)
Comment (zygoloid) [reply toolchain/sem_ir/type.h]: `AreEqual...` seems to work quite nicely with the value store accessors: `...types().AreEqual...`.

Done, and also renamed the corresponding function on the `constant_values` store.

File: toolchain/check/generic.cpp (hunk 7 above)
Comment (zygoloid) [reply toolchain/check/generic.cpp]: Hm, yeah, while the collection could get reallocated (eg, by import ref handling), the number of instructions at this level on the stack really shouldn't change. Made this change and also added a CHECK.

File: toolchain/check/subst.cpp (hunk 3 above)
Comment (zygoloid) [reply toolchain/check/subst.cpp:115]: Noted. I think this does look less weird in the full context of the file.

File: toolchain/check/subst.h (hunk 5 above)
Comment (zygoloid) [reply toolchain/check/subst.h]: Changed to talk more about what the return value means and less about what `SubstInst` will do with it.

File: toolchain/check/subst.h (hunk 6 above)
Comment (zygoloid) [reply toolchain/check/subst.h]: Did a pass over the comments.

**Review by zygoloid — COMMENTED** [review]
> The flow seems fine (although may be worth considering a class with virtual functions instead of lambda parameters?), but a pass on comments seems like it could be helpful.

Switched to a class with virtual functions as suggested. I think that is an improvement. I didn't find a really satisfying name for the class, but I think `SubstInstCallbacks` is at least clear as to its purpose.

**Review by jonmeow — APPROVED** [review]
Thanks! The edits are helpful for me.

As noted in chat, in my head I'd been thinking a full class for `SubstInst` logic, not just the callbacks, but this still seems a little easier to grasp.
